Print warnings when allowlist doesn't match anything

pvdrz opened the pull request:
This PR introduces a way to store warning messages inside `Bindings` so they can be emitted later. This allows `bindgen` to use `cargo:warnining=MESSAGE` in build scripts to show warnings and still produce warning log entries (which means this PR would fix https://github.com/rust-lang/rust-bindgen/issues/2167). 

We still have some questions about this that we would like to resolve before promoting this to a non-draft PR:
- [x] Should we emit log entries AND build scripts warnings at the same time? or should we favor one over the other?
  Apparently cargo won't print warnings for crates.io dependencies unless `-vv` is passed (https://github.com/rust-lang/cargo/issues/3777). I decided to float the warning messages all the way up and add a `Bindings::take_warnings` method so the user can print them directly in the build script. I also added a `print_warnings` macro to ease this process.

- [x] Is there any functionality in the test suite to capture stdout/stderr to be able to test that the warnings are being printed correctly?

cc @amanjeev

Edit: As pointed out by @kulp (thanks!), the warning log entries are still emitted..

kulp commented on src/ir/context.rs:
```diff
@@ -2430,17 +2437,27 @@ If you encounter an error missing from this list, please file an issue or a PR!"
         for item in self.options().allowlisted_functions.unmatched_items() {
-            warn!("unused option: --allowlist-function {}", item);
+            let warn = format!("unused option: --allowlist-function {}", item);
+            warn!("{}", warn);
```
I am just notating here that this PR does not stop emitting warnings the usual way (something that was not initially clear to me when reading the PR description).

kulp reviewed (approved):
This seems reasonable to me, and if @emilio does not see a problem maybe we can merge it soon. You could rebase it down to a few commits that all pass the tests, or we could squash it when merging.

pvdrz commented:
> This seems reasonable to me, and if @emilio does not see a problem maybe we can merge it soon. You could rebase it down to a few commits that all pass the tests, or we could squash it when merging.

Thanks for your review :) I addressed your comments and rebased the PR so all commits pass the tests individually

emilio commented on src/lib.rs:
```diff
@@ -2583,6 +2598,12 @@ impl Bindings {
+
+    /// Take all the warning messages.
+    #[inline]
+    pub fn take_warnings(&mut self) -> impl Iterator<Item = String> + '_ {
```
Can we just have `pub fn warnings(&self) -> &[String] { &self.warnings }` (effectively)?

I don't see why it's useful to take ownership of them.

pvdrz replied:
I thought it would be better to consume them so you don't have the chance to emit them twice.

emilio commented on src/lib.rs:
```diff
@@ -33,6 +33,19 @@ mod log_stubs;
+/// If you are using `bindgen` outside of a build script you should use [`Bindings::take_warnings`]
+/// directly instead.
+#[macro_export]
+macro_rules! print_warnings {
```
I also don't see why this needs to be a macro, can it just be a public method on `Bindings`?

pvdrz replied:
I don't have any way to test it. But apparently `crates.io` dependencies won't emit build script warnings unless `-vv` is passed to `cargo`. I don't know if functions from a dependency fall in this category and wrote a macro to fall into the safe side

pvdrz replied:
If we add a method for this we wouldn't actually need to have a `warnings` method.

emilio commented on src/ir/context.rs:
```diff
@@ -2430,17 +2437,27 @@ If you encounter an error missing from this list, please file an issue or a PR!"
-            warn!("unused option: --allowlist-type {}", item);
+            let warn = format!("unused option: --allowlist-type {}", item);
+            warn!("{}", warn);
+            warnings.push(warn);
```
Rather than duplicating all this, what about just changing:

```
warn!("unused option: ...");
```

with:

```
self.warnings.push(format!(...))
```

Then just call `warn!` at the end on a loop?

emilio reviewed (commented):
Looks sensible, but I think we can do a couple simplifications.

pvdrz commented:
@emilio I've addressed your comments. I'll squash them after your next review so you can still see the diff

emilio reviewed (approved):
Looks great, thanks!
